labels.go

```diff
@@ -0,0 +1,299 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"bytes"
+	"errors"
+	"sort"
+	"sync"
+
+	promproto "github.com/prometheus/client_model/go"
+)
+
+const (
+	// DefaultLabelName is used in place of empty label names.
+	DefaultLabelName = "default"
+
+	// DefaultLabelValue is used in place of empty label values.
+	DefaultLabelValue = "default"
+)
+
+var (
+	// Match the Prometheus error message.
+	errInconsistentCardinality = errors.New("inconsistent label cardinality")
+
+	_digesterPool = sync.Pool{New: func() interface{} {
+		return &digester{make([]byte, 0, 128)}
+	}}
+)
+
+// A digester creates a null-delimited byte slice from a series of variable
+// label values. It's an efficient way to create map keys from metric names and
+// labels.
+type digester struct {
+	bs []byte
+}
+
+// For optimal performance, be sure to free each digester.
+func newDigester() *digester {
+	d := _digesterPool.Get().(*digester)
+	d.bs = d.bs[:0]
+	return d
+}
+
+func (d *digester) add(prefix, s string) {
+	if len(d.bs) > 0 {
+		// separate labels with a null byte
+		d.bs = append(d.bs, '\x00')
+	}
+	d.bs = append(d.bs, prefix...)
+	d.bs = append(d.bs, s...)
+}
+
+func (d *digester) digest() []byte {
+	return d.bs
+}
+
+func (d *digester) free() {
+	_digesterPool.Put(d)
+}
+
+// Labels describe the dimensions of a metric.
+type Labels map[string]string
+
+func zip(pairs []*promproto.LabelPair) Labels {
+	ls := make(Labels, len(pairs))
+	for _, pair := range pairs {
+		if pair == nil || pair.Name == nil || pair.Value == nil {
+			continue
+		}
+		ls[*pair.Name] = *pair.Value
+	}
+	return ls
+}
+
+// less provides a stable ordering among label sets.
+func (l Labels) less(other Labels) bool {
+	leftNames := make([]string, 0, len(l))
+	for k := range l {
+		leftNames = append(leftNames, k)
+	}
+	sort.Strings(leftNames)
+	left := newDigester()
+	for _, n := range leftNames {
+		left.add("", n)
+		left.add("", l[n])
+	}
+
+	rightNames := make([]string, 0, len(other))
```

Comment:
can we pull out the logic to build a digest out of labels so it's not duplicate between left and right

Reply:
Good point.